src/librustc_data_structures/interner.rs

```diff
@@ -0,0 +1,291 @@
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::marker::PhantomData;
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::hash::BuildHasher;
+use std::mem::{self, size_of};
+use std::ptr::{Unique, NonNull};
+use std::alloc::{Global, Alloc};
+use std::collections::hash_map::RandomState;
+use std::borrow::Borrow;
+use std::fmt;
+
+const ENTRIES_PER_GROUP: usize = 5;
+
+#[repr(align(64), C)]
+pub struct Group {
+    hashes: [u32; ENTRIES_PER_GROUP],
+    size: u32,
+    values: [u64; ENTRIES_PER_GROUP],
+}
+
+impl Group {
+    #[inline(always)]
+    fn search_for_empty(&self) -> Option<usize> {
+        if self.size != ENTRIES_PER_GROUP as u32 {
+            Some(self.size as usize)
+        } else {
+            None
+        }
+    }
+
+    #[inline(always)]
+    fn search_with<K, F: FnMut(&K) -> bool>(&self, eq: &mut F, hash: u32) -> Option<(usize, bool)> {
+        for i in 0..ENTRIES_PER_GROUP {
+            let h = unsafe { *self.hashes.get_unchecked(i) };
+            if h == hash && eq(unsafe { mem::transmute(self.values.get_unchecked(i)) }) {
+                return Some((i, false))
+            }
+        }
+        self.search_for_empty().map(|i| (i, true))
+    }
+
+    #[inline(always)]
+    fn set(&mut self, pos: usize, hash: u32, value: u64) {
+        unsafe {
+            *self.hashes.get_unchecked_mut(pos) = hash;
+            *self.values.get_unchecked_mut(pos) = value;
+        }
+    }
+
+    #[inline(always)]
+    fn iter<F: FnMut(u32, u64)>(&self, f: &mut F) {
+        for i in 0..ENTRIES_PER_GROUP {
+            unsafe {
+                let h = *self.hashes.get_unchecked(i);
+                if h != 0 {
+                    f(h, *self.values.get_unchecked(i))
+                }
+            }
+        }
+    }
+}
+
+pub struct Table {
+    group_mask: usize,
+    size: usize,
+    capacity: usize,
+    groups: Unique<Group>,
+}
+
+pub struct RawEntry {
+    group: *mut Group,
+    pos: usize,
+    empty: bool
+}
+
+impl Drop for Table {
+    fn drop(&mut self) {
+        if self.group_mask == 0 {
+            return;
+        }
+
+        unsafe {
+            Global.dealloc_array(
+                NonNull::new_unchecked(self.groups.as_ptr()),
+                self.group_mask + 1
+            ).unwrap();
+        }
+    }
+}
+
+impl Table {
+    fn new(group_count: usize) -> Table {
+        assert!(size_of::<Group>() == 64);
+        let groups: NonNull<Group> = Global.alloc_array(group_count).unwrap();
+        let capacity2 = group_count * ENTRIES_PER_GROUP;
+        let capacity1 = capacity2 - 1;
+        //let capacity = (capacity1 * 10 + 10 - 1) / 11;
+        let capacity = (capacity1 * 10 + 10 - 1) / 13;
+        //println!("capacity1 {} capacity {}", capacity1, capacity);
+        assert!(capacity < capacity2);
```

Comment:
I don't quite understand this capacity calculation. It seems to be adapted from what the std hashmap does but it could need better variable names and comments explaining the reasoning behind it.